tokio/src/sync/semaphore_ll.rs

```diff
@@ -103,46 +96,55 @@ struct WaiterNode {
 struct SemState(usize);
 
 /// Permit state
-#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+#[derive(Debug, Copy, Clone)]
 enum PermitState {
-    /// The permit has not been requested.
-    Idle,
-
-    /// Currently waiting for a permit to be made available and assigned to the
+    /// Currently waiting for permits to be made available and assigned to the
     /// waiter.
-    Waiting,
+    Waiting(u16),
 
-    /// The permit has been acquired.
-    Acquired,
+    /// The number of acquired permits
+    Acquired(u16),
 }
 
-/// Waiter node state
-#[derive(Debug, Copy, Clone, Eq, PartialEq)]
-#[repr(usize)]
-enum NodeState {
-    /// Not waiting for a permit and the node is not in the wait queue.
-    ///
-    /// This is the initial state.
-    Idle = 0,
+/// State for an individual waker node
+#[derive(Debug, Copy, Clone)]
+struct WaiterState(usize);
 
-    /// Not waiting for a permit but the node is in the wait queue.
-    ///
-    /// This happens when the waiter has previously requested a permit, but has
-    /// since canceled the request. The node cannot be removed by the waiter, so
-    /// this state informs the receiver to skip the node when it pops it from
-    /// the wait queue.
-    Queued = 1,
+/// Waiter node is in the semaphore queue
+const QUEUED: usize = 0b001;
 
-    /// Waiting for a permit and the node is in the wait queue.
-    QueuedWaiting = 2,
+/// Semaphore has been closed, no more permits will be issued.
+const CLOSED: usize = 0b10;
 
-    /// The waiter has been assigned a permit and the node has been removed from
-    /// the queue.
-    Assigned = 3,
+/// The permit that owns the `Waiter` dropped.
+const DROPPED: usize = 0b100;
 
-    /// The semaphore has been closed. No more permits will be issued.
-    Closed = 4,
-}
+/// Represents "one requested permit" in the waiter state
+const PERMIT_ONE: usize = 0b1000;
+
+/// Masks the waiter state to only contain bits tracking number of requested
+/// permits.
+const PERMIT_MASK: usize = usize::MAX - (PERMIT_ONE - 1);
+
+/// How much to shift a permit count to pack it into the waker state
+const PERMIT_SHIFT: u32 = PERMIT_ONE.trailing_zeros();
```

Comment:
nit, take it or leave it: i wonder if the bit-packing stuff in https://github.com/tokio-rs/tokio/blob/master/tokio/src/util/bit.rs can/should be used here?